### PR TITLE
Issue 1349 - osc project supports switching by context name

### PR DESCRIPTION
**Pull request by fabianofranz**
Fixes https://github.com/openshift/origin/issues/1349

### CODE REVIEW

**Comment by fabianofranz** [comment]
@smarterclayton Take a look. 
One note: I'm displaying context name AND project name, but not server and user. From the config file I don't know which username I am, one idea is to fetch it with a call to `Users().Get(~)`, but it would make `osc project <context>` call the server. And I think just printing the `authInfo` name does not make much sense (`authInfo.name` *can* be the username, but not necessarily will be). 
Any preference?

File: pkg/cmd/cli/cmd/project.go [hunk 1]
```diff
@@ -43,71 +52,77 @@ func NewCmdProject(f *clientcmd.Factory, out io.Writer) *cobra.Command {
...
 					if err != nil {
-						if errors.IsNotFound(err) {
+						if errors.IsNotFound(err) || clientcmd.IsForbidden(err) {
 							glog.Fatalf("The project '%v' specified in your config does not exist or you do not have rights to view it.", currentProject)
 						}
 						checkErr(err)
 					}
 
-					fmt.Printf("Using project '%v'.\n", currentProject)
+					fmt.Printf("Using context '%v' with project '%v'.\n", rawCfg.CurrentContext, currentProject)
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/project.go]: Wasn't sure what your comment meant, but make this an if block "if currentProject != currentContext { Using project %v from context named %v }   else { Using project %v }"

File: pkg/cmd/cli/cmd/project.go [hunk 2]
```diff
@@ -115,8 +130,68 @@ func NewCmdProject(f *clientcmd.Factory, out io.Writer) *cobra.Command {
...
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return configStore, err
+}
+
+func clusterAndAuthEquality(clientCfg *kclient.Config, cluster clientcmdapi.Cluster, authInfo clientcmdapi.AuthInfo) bool {
+	return cluster.Server == clientCfg.Host &&
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/project.go]: This could have changed because the user specified it on the CLI?

File: pkg/cmd/cli/cmd/project.go [hunk 3]
```diff
@@ -19,7 +27,7 @@ func NewCmdProject(f *clientcmd.Factory, out io.Writer) *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "project <project-name>",
 		Short: "switch to another project",
-		Long:  `Switch to another project and make it the default in your configuration.`,
+		Long:  `Switch to another project or context and make it the default in your configuration.`,
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/project.go]: Just say "project".  We'll define what contexts are when the user hits them.  Also, this should be a longer long description.

File: pkg/cmd/cli/cmd/project.go [hunk 4]
```diff
@@ -115,8 +130,68 @@ func NewCmdProject(f *clientcmd.Factory, out io.Writer) *cobra.Command {
...
+			return nil, fmt.Errorf(msg)
+		}
+
+		return nil, err
+	}
+
+	return project, nil
+}
+
+func getProjectsNames(oClient *client.Client) ([]string, error) {
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/project.go]: getProjectNames

File: pkg/cmd/cli/cmd/project.go [hunk 5]
```diff
@@ -115,8 +130,68 @@ func NewCmdProject(f *clientcmd.Factory, out io.Writer) *cobra.Command {
...
+
+func getProject(oClient *client.Client, name string) (*api.Project, error) {
+	project, err := oClient.Projects().Get(name)
+
+	if err != nil {
+		if errors.IsNotFound(err) || clientcmd.IsForbidden(err) {
+			msg := fmt.Sprintf("A project named '%v' does not exist or you do not have rights to view it.", name)
+
+			if projectsNames, err := getProjectsNames(oClient); err != nil {
+				msg += fmt.Sprintf(" Your projects are: %v.", strings.Join(projectsNames, ", "))
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/project.go]: One per line

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/project.go]: Also, "on this server".  Then, at the end of the message, if they have a different server in their config, display a message: "To see projects on another server, pass '--server='"

**Comment by smarterclayton** [comment]
If they provided a username show it, otherwise don't.  We may want to store that info in the context eventually but for now it's fine to not have it.

----- Original Message -----

File: pkg/cmd/cli/cmd/project.go (hunk 1 above)
Comment (fabianofranz) [inline pkg/cmd/cli/cmd/project.go]: Fixed.

File: pkg/cmd/cli/cmd/project.go (hunk 2 above)
Comment (fabianofranz) [inline pkg/cmd/cli/cmd/project.go]: Yes.

File: pkg/cmd/cli/cmd/project.go (hunk 3 above)
Comment (fabianofranz) [inline pkg/cmd/cli/cmd/project.go]: Fixed (not the longer long desc).

File: pkg/cmd/cli/cmd/project.go (hunk 4 above)
Comment (fabianofranz) [inline pkg/cmd/cli/cmd/project.go]: Fixed, I actually repurposed this method.

File: pkg/cmd/cli/cmd/project.go (hunk 5 above)
Comment (fabianofranz) [inline pkg/cmd/cli/cmd/project.go]: Done.

**Comment by fabianofranz** [comment]
@smarterclayton All comments addressed. I'm showing the server, but never the username - they are not supposed to provide `--username` to this command (I have a TODO to de-register `--username` and `--password` for every command other than `login`).

**Comment by fabianofranz** [comment]
AND looks like stuff changed and I have to rebase.  :)

**Comment by fabianofranz** [comment]
Rebased.

**Comment by fabianofranz** [comment]
@smarterclayton Fixed the latest comments.

**Comment by smarterclayton** [comment]
Let me try it out.

**Comment by smarterclayton** [comment]
```
$ osc project foo
Using server "https://localhost:8443".
Now using project "foo" from context named "b4596c40-d1b5-11e4-b7f8-7831c1b76042".
```

Should be using "foo" as the context name (by default) and the message should be

"Now using project "foo" on server https://localhost:8443"

because context name == project name.  Don't need two lines for server.

----- Original Message -----

**Comment by smarterclayton** [comment]
```
$ osc project foo
Using server "https://localhost:8443".
Now using project "foo" from context named "b4596c40-d1b5-11e4-b7f8-7831c1b76042".

$ osc project bar
Using server "https://localhost:8443".
Now using project "bar" from context named "fc10fe66-d1b5-11e4-b654-7831c1b76042".

$ osc project foo
Using server "https://localhost:8443".
Now using project "foo" from context named "fd1e107a-d1b5-11e4-9b48-7831c1b76042".
```

The last one should be using the context name from the first time (which should be foo, but it looks like it's generating a new one every time)

1. Clean slate (no contexts)
2. osc project foo
3. Have one context "foo" pointing to project "foo"
4. osc project bar
5. Have two contexts "foo" and "bar" pointing to those projects
6. osc project bar --server=<server>
7. should create a new context "bar-<servername>" pointing to project bar on that server (if i have a single cluster with that server name)
8. osc project bar --cluster=<clustername>
9. should create a new context "bar-<clustername>" pointing to project bar on that server
10. osc project bar --context=<contextname>
11. if context name does not exist, it should create it pointing to that project.  otherwise, if the context already exists pointing to a different project and different server, it should give me an error.

I don't think you need to implement 6-9 here, but eventually we'll need to be able to do that.

----- Original Message -----

**Comment by fabianofranz** [comment]
@smarterclayton Most issues have been addressed, please give it a try.

I also integrated the context identifier generator between `project` and `login` that will now use the same logic. Will try to generate the best identifier possible with the information available, removing unwanted characters like `:./`. Will result into something like this (in order, subsequent ones will be used when previous ones are taken):

```
foo
foo-localhost
foo-localhost-8443
foo-localhost-8443-username
foo-localhost-8443-username-0
foo-localhost-8443-username-1
(...)
UUID if nothing else works
```

**Comment by smarterclayton** [comment]
When would uid be necessary?

**Comment by fabianofranz** [comment]
Most likely never, but if every other 104 context identifiers we tried to generate are taken, we need something to be the context identifier.

----- Original Message -----

**Comment by smarterclayton** [comment]
I think in that case it would be ok to bail and report an error to the user that they need to provide a context name on the CLI

----- Original Message -----

**Comment by smarterclayton** [comment]
LGTM [merge] - follow ups opened as issues.

**Comment by fabianofranz** [comment]
@smarterclayton Just fixed your latest comment.

**Pull request merged**